### ISSUE TITLE
attrs: [Errno 28] No space left on device: '<FD x>'

### ISSUE THREAD

**Issue by Thuffir** [borgbackup/borg#3808]
## Have you checked borgbackup docs, FAQ, and open Github issues?
Yes

## Is this a BUG / ISSUE report or a QUESTION?
Not sure. Probably just a question.

#### Your borg version (borg -V).
    borg 1.1.5

#### Operating system (distribution) and version.
    Linux 4.9.0-6-amd64 #1 SMP Debian 4.9.82-1+deb9u2 (2018-02-21) x86_64 GNU/Linux
    ntfs-3g 2016.2.22AR.1 integrated FUSE 28

#### Hardware / network configuration, and filesystems used.
ext4 / ntfs

#### How much data is handled by borg?
approx 500 GB

#### Full borg commandline that lead to the problem (leave away excludes and passwords)
    borg extract ::archivename

## Describe the problem you're observing.
I did a full backup of an NTFS drive (mounted on linux over ntfs-3g in read only mode).

When I try to restore the files to an ext4 filesystem I get the following errors for some files:

    attrs: [Errno 28] No space left on device: '<FD xx>'

Interestingly the files I get those errors for are always picture files (BMP, TIFF,  JPG etc).

So I guess this might be some (x)attr incompatibility / limitation between NTFS and ext4 and nothing to worry about right?

**Comment by ThomasWaldmann** [comment]
I remember we already had a similar issue before. In that old case the strange error msg was caused by trying to restore ntfs fs special files/directories (stuff that rather should have been excluded from the backup).

In your case, the "attrs" means it had problems to restore some attributes (uid, gid, mode, timestamps, acls, xattrs, bsdflags). For a lot of these, we catch all OSErrors, so what's left is mode, acls and xattrs.

For xattrs, we catch some OSError, but not "no space left" (because that should mean that the disk is full and we can not just ignore that).

The relevant code is there:
https://github.com/borgbackup/borg/blob/1.1.5/src/borg/archive.py#L659

**Comment by ThomasWaldmann** [comment]
It would be interesting to find more about this, what precisely is causing the issue.

So maybe do an extraction with `--list` and post the line with the error plus 2 lines before it and 2 lines after it. If there is any traceback, post the full traceback.

**Comment by Thuffir** [comment]
I managed to capture this issue in a test archive:
[borg_issue_3808.tar.gz](https://github.com/borgbackup/borg/files/1983056/borg_issue_3808.tar.gz)

It produces (at least on my system) the mentioned error message when I try to extract it on an ext4 filesystem:
```
root@xxx# borg extract ../test::test
test: attrs: [Errno 28] No space left on device: '<FD 6>'
```

**Comment by Thuffir** [comment]
`getfattr -d` shows some strange xattrs hanging on that file. Maybe it's something ext4 can't cope with.

**Comment by ThomasWaldmann** [comment]
With your test repo and slightly modified borg code:
```
$ borg extract --list ../test::test
test
/home/tw/Desktop/test2/test: No space left while setting extended attribute user.Q30lsldxJoudresxAaaqpcawXc
```

WTF?

**Comment by ThomasWaldmann** [comment]
@Thuffir try the patch in #3812.

**Comment by Thuffir** [comment]
Actually borg did not crash on this error but continued to extract the next file(s). Is this behavior still the same now?

**Comment by ThomasWaldmann** [comment]
Well, for not specially treated OSErrors, it just re-raised them there (before my change).

Some upstream exception handler then skipped to the next file (because it is an error of an individual restored file, not a fatal error in the repository or other borg on-disk data structures).

Now (after my change), ENOSPC gets special treatment: log a warning, set return code to 1 (WARNING) and just continue with next xattr key/value pair.

**Comment by Thuffir** [comment]
Perfect.
So as it was not a borg bug but rather some strange filesystem xattr interoperability problem I would close this issue.

**Comment by ThomasWaldmann** [comment]
It will be automatically closed once the code is merged into master (and also other branches).

**Comment by ThomasWaldmann** [comment]
The root cause is that (older) ext4 implementations only can deal with 4kiB xattrs.

And the ext4 code really returns ENOSPC for that case (and not E2BIG).

**Comment by ThomasWaldmann** [comment]
BTW, can you find out what's the usage or origin of that strange-looking xattr you have there?

**Comment by ThomasWaldmann** [comment]
Tagging as a bug.

I guess at least the not-applying of any subsequent xattrs (and other attrs set afterwards in the same method) is a bug in our code.

Whether ENOSPC (or E2BIG) is the correct error code (and whether that is a bug in ext4) I can not say, but even if it were, we have to live with that.

**Comment by enkore** [comment]
ENOSPC is correct.
E2BIG is for reading an xattr exceeding the VFS capabilities (which on Linux is 64K).
